contentpacks/khanacademy.py

```diff
@@ -194,74 +196,258 @@ def retrieve_translations(crowdin_project_name, crowdin_secret_key, lang_code="e
+# 'Time value of money' at http://s3.amazonaws.com/KA-youtube-converted/733mgqrzNKs.mp4/733mgqrzNKs.mp4...
+# 'Applying the metric system' at http://s3.amazonaws.com/KA-youtube-converted/CDvPPsB3nEM.mp4/CDvPPsB3nEM.mp4...
+# 'Mixed numbers: changing to improper fractions' at http://s3.amazonaws.com/KA-youtube-converted/xkg7370cpjs.mp4/xkg7370cpjs.mp4...
+
+slug_key = {
+    "Topic": "slug",
+    "Video": "readable_id",
+    "Exercise": "name",
+}
+
+def modify_slugs(nodes) -> list:
+    for node in nodes:
+        node["slug"] = node.get(slug_key.get(node.get("kind")))
+    return nodes
+
+id_key = {
+    "Topic": "slug",
+    "Video": "youtube_id",
+    "Exercise": "name",
+}
+
+def modify_ids(nodes) -> list:
+    for node in nodes:
+        node["id"] = node.get(id_key.get(node.get("kind")))
+    return nodes
+
+def apply_black_list(nodes) -> list:
+    return [node for node in nodes if node.get("slug") not in slug_blacklist]
+
+def group_by_slug(count_dict, item):
+    # Build a dictionary, keyed by slug, of items that share that slug
+    if item.get("slug") in count_dict:
+        count_dict[item.get("slug")].append(item)
+    else:
+        count_dict[item.get("slug")] = [item]
+    return count_dict
+
+def create_paths_remove_orphans_and_empty_topics(nodes) -> list:
+    node_dict = {node.get("id"): node for node in nodes}
+    # Set the slug on the root node to "khan"
+    node_dict["x00000000"]["slug"] = "khan"
+
+    node_list = []
+
+    def recurse_nodes(node, parent_path=""):
+
+        """
+        :param node: dict
+        :param parent_path: str
+        """
+        node["path"] = parent_path + node.get("slug") + "/"
 
+        children = node.pop("child_data", [])
 
-def retrieve_kalite_exercise_data(url=None, force=False) -> dict:
+        if children:
+            children = [node_dict.get(child.get("id")) for child in children if node_dict.get(child.get("id"))]
+
+            counts = reduce(group_by_slug, children, {})
+            for items in counts.values():
+                # Slug has more than one item!
+                if len(items) > 1:
+                    i = 1
+                    # Rename the items
+                    for item in items:
+                        if item.get("kind") != "Video":
+                            # Don't change video slugs, as that will break internal links from KA.
+                            item["slug"] = item["slug"] + "_{i}".format(i=i)
+                            item["path"] = node.get("path") + item["slug"] + "/"
+                            i += 1
+
+        for child in children:
+            recurse_nodes(child, node.get("path"))
+
+        if children or node.get("kind") != "Topic":
+            node_list.append(node)
+
+    recurse_nodes(node_dict["x00000000"])
+
+    return node_list
+
+@cache_file
+def download_and_clean_kalite_data(url, path) -> str:
+    data = requests.get(url)
+    while data.status_code != 200:
```

Comment:
Is this apt to fail occassionally, hence the for loop? This might cause an infinite loop when the url is, say, pointing to an invalid location (404). Might want to add a set # of loops for this.